### ISSUE TITLE
Subfolders of moved folders not synced

### ISSUE THREAD

**Issue by eboth** [nextcloud/desktop#1000]
### Expected behaviour
When on your local Windows system, you move a folder into your Nextcloud folder within the same NTFS partition (did't test other file systems), the Nextcloud client should recursively synchronize all directory depths, i.e. all files and folders in all nested subfolders.

### Actual behaviour
Moved folders are only synchronized up to the first level. 
This is the directory `test` as copied into the Nextcloud folder:
```
+-test/
  |
  +-level1/
  | |
  | +-level2/
  | |
  | +-level2.txt
  |
  +-level1.txt
```
This is the directory structure as copied by the Nextcloud client to the server:
```
+-test/
  |
  +-level1/
  |
  +-level1.txt
```
### Steps to reproduce
1. Create a directory `test` with at least one subfolder `level1` *outside* your Nextcloud folder hierarchy, but on the same partition as your Nextcloud folder.
2. Place content (files/folders) in `level1`.
3. Move (not copy) `test` into your Nextcloud folder (Windows default `C:\Users\username\Nextcloud`).
4. Wait for Nextcloud client to finish synchronization.
5. Check contents of `test` on the server; confirm that the subfolder `level1` is empty on the server.

### Client configuration
2.5.1final (build 20181204)

Operating system:
Windows 7 SP 1 (6.1)

File system:
NTFS

Installation path of client:
c:\Program Files (x86)\Nextcloud\

### Server configuration
13.0.8.2

### Logs
1. Client logfile: 
https://gist.github.com/eboth/49d061cd66724e3e9dc699cfc4d1591f

While the log file was captured, the folder `test_copied` (with two levels of subdirectories as shown in the directory listing above) was copied to the Nextcloud folder, and the folter `test_moved` (same contents) was moved to the Nextcloud folder. The log seems to show that in the copied folder, all nodes were recognized correctly, while in the moved folder, the Nextcloud client never became aware of the nodes below `level1`.

**Comment by packi** [comment]
I can confirm that on Linux, script to create a test folder structure can be found here: https://gist.github.com/packi/1361912055bd03490bbb6a8906d9800e

I've initially created a bug for that on the owncloud client repos with more details as I didn't realize there was a separate repos for Nextcloud. https://github.com/owncloud/client/issues/6949

**Comment by abysso2** [comment]
Same here with linux client 2.5.1 ans server 15.0.4 ...

**Comment by dude2k5** [comment]
Same here, have the same issue. Main folder syncs, but subfolder did not.

**Comment by estepix** [comment]
Same problem here:

Client

- Version: 2.5.1final (build 20181204)
- OS: Ubuntu 16.04.5 LTS xenial

Server

- Versions: 14 and 15.0.4
- OS: Ubuntu 16.04.5 LTS
- Web Server: Nginx 1.15.8-8.6.0.1

Using PPA:

- http://ppa.launchpad.net/nextcloud-devs/client/ubuntu

Packages version:
```
ii  libnextcloudsync0                                    2.5.1-20181204.111806~xenial1                                               amd64        Nextcloud sync library
ii  nextcloud-client                                     2.5.1-20181204.111806~xenial1                                               amd64        Nextcloud desktop sync client
ii  nextcloud-client-l10n                                2.5.1-20181204.111806~xenial1                                               all          Nextcloud client internatialization files
```

Log shows that _csync_detect_update only detects first level:
```
[_csync_detect_update 	Checking for rename based on inode # 11010575
[_csync_detect_update 	file: nextcloud/myclient/test, instruction: INSTRUCTION_NEW <<=
[_csync_detect_update 	Checking for rename based on inode # 11010705
[_csync_detect_update 	file: nextcloud/myclient/test/level1, instruction: INSTRUCTION_NEW <<=
```

**Comment by hkiang01** [comment]
workaround: quit and restart the desktop client. Tested with the `nextcloud-client` Ubuntu ppa.

**Comment by estepix** [comment]
I have read about that "workdaround" before and also tested and yes, it works, however restarting the client every time we add a few folders to nextcloud does not seem viable on a daily work basis

**Comment by eboth** [comment]
@hkiang01, thanks! Confirmed on Windows. This makes the bug a lot less lethal as you can hope that sooner or later the files will end up on the server.

**Comment by SonRiab** [comment]
Possible duplicate #1083?

**Comment by bjoernv** [comment]
I can not reproduce this. All files from file structure https://github.com/nextcloud/desktop/issues/1000#issuecomment-453500896 are synchronized.

Client

    * Version: 2.5.1

    * OS: openSUSE Tumbleweed 20190225

    * File system: ext4


Server

    * Version: 14.0.7

    * OS: Ubuntu 14.0.4 LTS

    * Web Server: Apache 2.4.7

    * PHP 7.2

**Comment by aaronpk** [comment]
Same problem here on MacOS. I was trying to migrate files from Dropbox and noticed that any time I dragged a folder with lots of subfolders over, none of the subfolders had any contents synced.

Confirmed that quitting and re-launching the desktop client causes a complete sync.

**Comment by dnv** [comment]
I can confirm this happening to me on Windows 10 with Nextcloud sync client 2.5.1final (build 20181204). Let's just say it made me quite vary of using Nextcloud to begin with.

**Comment by driph** [comment]
Confirming the same issue,  2.5.1final (build 20181204) on macOS 10.13.6.
Restarted the client resolved it for the bugged sync, but any further additions of nested directories after the sync on start result in the same behavior and workaround.

**Comment by wpoet** [comment]
Confirming the inability of the client to sync all subfolders and files: Ubuntu 18.10, Version 2.5.1final (build 20181204). This is really a killerbug. Hopefully this will be fixed soon. We are trying to introduce Nextcloud to students of our school, but this is makes nextcloud a real no go.

Does anyone on this forum know a better work around than restarting the desktop client? Maybe a command on the terminal?

**Comment by abysso2** [comment]
I just upgraded to the linux-client 2.5.2, still the same issue. I can not believe it. The server ist such an excellent piece of software, but the client really can not be trusted. That is a real showstopper - why does the owncloud client work as expected?

**Comment by adrianriedo** [comment]
Spent a full day looking on Windows 10 machine what I made wrong (because I did not see expected files on the Server) and finally found this issue.
I can confirm the problem (moving folder structure, sync not done 100%).
This is a showstopper to move all our data from Seafile (which has a perfect working client) to Nextcloud.
The Server Side of Nextcloud is quite nice - but the very most fundamental thing (sync files with the Desktop Client) can not be trusted all the times...
I hope someone who is experienced with the Desktop Client can correct this.

**Comment by dothebakedpotato** [comment]
Also have this issue (with Linux desktops/laptops). Can occur when copying directories as well as moving. 

Unfortunately Nextcloud is not fit for enterprise or business without a working and robust client. It makes me extremely nervous about the prospect of losing critical data.

**Comment by ziggr** [comment]
I just got bit by this bug, too.

Client 2.5.2git

```
$ uname -a
Linux gemini 4.15.0-47-generic #50~16.04.1-Ubuntu SMP Fri Mar 15 16:06:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ mount | grep nvme
/dev/nvme0n1p2 on / type ext4 (rw,noatime,discard,errors=remount-ro,data=ordered)
```
A whole .git repo and its subdirectories, along with the contents of a lot of subdirectories of other files, all missing from server. Upon restarting the desktop client, the files synced to server (phew!).

This is really heartbreaking after spending a whole weekend creating this server and getting everything to work. How can I use Nextcloud to sync my source code if my projects all have subdirectories (and git repos!).

**Comment by adrianriedo** [comment]
I really hope someone who codes for the desktop client reads this and will improve it.
Fact is that we used Owncloud some years ago and had very similar issues.
Then we used Seafile for the last 2-3 years and it worked fine with its desktop client.
Nextcloud has quite a momentum right now, so we thought to give it a try again - unfortunately the desktop sync is still a problem as it seems - I can not tell all users to close and restart the client every day to make shure the files are synced...

Nextcloud has some nice features on the Server - but when it comes to syncing with desktop client it seems to have real issues.
Can it really be that so many users use it without even noticing that files are missing on the server?
Strange

**Comment by Kambrian** [comment]
I cannot believe such a HUGE MAJOR bug exists with this software and not get fixed (or even noticed?) for such a long time till today. 

I am regretting the efforts I spent trying to set up my personal cloud with nextcloud. This should not be anything from a commercial company. Just hard to believe.

**Comment by Germano0** [comment]
Please retweet https://twitter.com/GermanoMassullo/status/1125077393272655872

**Comment by adrianriedo** [comment]
Worked with Nextcloud now for 2 weeks after some years of Seafile.
Now after this (and some other bugs in the Desktop Client that we found) we had to move back to Seafile again. The Desktop Client has also problems with multiple Sync Directories when you use Standby on your Computer - after wakeup syncing does not start on all Sync Folder (if you have more than a hand full of them).
The Web UI is nice and fancy - but syncing with Desktop is sadly a risk.
I would have loved to move from Seafile to Nextcloud - but not at this stage...

**Comment by jospoortvliet** [comment]
I can reproduce this.

**Comment by bill-mcgonigle** [comment]
>  most users do restart their computers once a week if not more often

This might be true on Windows, but I have hard enough time getting linux users to reboot for kernel security updates - uptimes over 90 days are quite typical.

**Comment by dnv** [comment]
This isn't true on Windows either, at least in a business environment
uptimes of 3+ weeks are more common than not.

The way this issue has been handled so far is a total joke. I am in the
process of rebuilding my server this very moment and will now be going with
Seafile instead. Oh, and telling everyone I know exactly how Nextcloud
manages it's bugfixing.

**Comment by kleinph** [comment]
I agree with @dnv, the handling of this bug looks unprofessional. There are two issues open (this and #876) since at least end of November and there seems to be no response from anyone of Nextcloud.

**Comment by dothebakedpotato** [comment]
Assuming that users will restart their computers is a bad excuse, and not reasonable in many environments. I have workstation uptimes on the order of months.

I need users' files to be synced reliably onto ZFS file servers. The ZFS pools are then backed up conventionally. This is a bug that is occurring regularly, many times a week. If I can't rely on the sync client then data loss is a very real and worrying possibility. 

The way this has been handled by Nextcloud is faintly ridiculous. It is not showing itself to be a serious or competent company.

**Comment by jospoortvliet** [comment]
@dothebakedpotato you are not a customer of us, are you? So we have no SLA or any other contract in place?

I believe that serious companies run software with a SLA from a vendor. Not doing that is indeed risking user data, and if my IT team would do that without telling me, I would not be happy.

**Comment by dnv** [comment]
> @dothebakedpotato you are not a customer of us, are you? So we have no SLA or any other contract in place?
> 
> I believe that serious companies run software with a SLA from a vendor. Not doing that is indeed risking user data, and if my IT team would do that without telling me, I would not be happy.

Public relations tip: you are not making how this looks for Nextcloud any better with this post.

Have you provided a fix for this issue to those who actually ARE paying you?

**Comment by jospoortvliet** [comment]
@dnv that might very well be. I will lock this now for contributors, this discussion is pointless. If you suffer from this as home user - please, restart your client regularly until we or any volunteer has had time to fix it. I'm sorry for this issue, we're working as much as we can but we have to serve our customers first.

If this bites you as enterprise user, contact our sales team.

**Comment by jancborchardt** [comment]
I’d like to remind everyone that we have a [Code of Conduct](https://nextcloud.com/code-of-conduct/).

- 🔥 Aggressive behavior does not make anyone more likely to fix anything. In fact, it makes it _less_ likely.
- 🤖 The people working on this project are, in fact, real people. Be friendly.
- 💸 Rather than wasting time writing an inflammatory comment, donate to fix this issue [on Bountysource](https://www.bountysource.com/issues/68330828-subfolders-of-moved-folders-not-synced).
- :heart: Remember that the software is fully free & open source. Treat the people working on it with respect.

Also thanks to everyone who has been constructive in this issue.

**Comment by jancborchardt** [comment]
Also, you might have seen that we are looking for a C++/Qt developer to join our desktop client team. Please spread the word: https://twitter.com/camiladotcodes/status/1115523047031808001 – that’s another thing you can do to actually help with this issue.

:heart:

**Comment by camilasan** [comment]
The reply about being a customer vs community might have come out wrong - and I am sorry for that - but please be assured that we are working on this issue. You might not get a reply but it doesn't mean I am not aware of the issue. I am actively looking to hire an additional person to join me on the desktop client. Please be patient and understand that we are very busy and doing everything we can to make Nextcloud work perfectly. Comments like the one we replied to can be hurtful when you're doing your best - please keep the discussion polite.

**Comment by jancborchardt** [comment]
And _huge_ thanks to @joshua-sterner for contributing a fix! Pull request is open at https://github.com/nextcloud/desktop/pull/1245 – everyone please help testing and review! 😊

**Comment by camilasan** [comment]
The PR #1245 works for me. Could you all run https://transfer.sh/X46OK/Nextcloud-PR-1245-bd9fcd88d85f48889e5e6fd55456cec7b123a22c-x86_64.AppImage to verify that too? Thanks.

**Comment by dothebakedpotato** [comment]
> @dothebakedpotato your company is not a customer of us, are you? So we have no SLA or any other contract in place with your company?
> 
> I believe that serious companies run software with a SLA from a vendor. Not doing that is indeed risking user data, and if my IT team would do that without telling me, I would not be happy.
> 
> Please note that github is for contributors to coordinate their work. We handle bugs from home users and plan our work. If your company uses Nextcloud and you look or ask for help here, you're in the wrong place. Customers can contact our team through https://portal.nextcloud.com and there they get prompt help.
> 
> (Edit: I updated the comment a few times to clarify that I'm clearly talking to a paid employee from another company who expects free support services. We're very happy to help home users with Nextcloud and we want to fix this bug, but having employees from other companies using their paid time to tell us we're not working fast enough for them for free is a little disappointing.)

You're making an awful lot of assumptions. Let me explain the situation. 

We're a mid enterprise of about 300 users. We have worked hard over a long time to build our business to where we are today and we are careful about how we spend our money. I have been looking for a "cloud" technology to replace our NFS/Kerberos file shares. Nextcloud looked most promising and so, since you release your product under the AGPL, I decided to try it out. I have been running a trial with a very small number of users to see whether it is a viable option. If it was suitable, of course we would buy an SLA.

Unfortunately for both of us (given the effort and time I have invested in trialling it), my conclusion is that it is not suitable. Your desktop client is buggy and puts data at risk. This is a long-standing bug and clearly the client is not your priority. I posted on this ticket to see what would happen. What happened is that I was told it was not a real issue and data was not at risk. I was attacked and then the ticket was locked so I could not give any response. My advice would be not to tell users that they are wrong: they know their use-cases better that you do. It is also not reasonable to demand that people buy an SLA before you will fix critical bugs.. I do not buy SLA's so that I can ask people to fix buggy software. It's a very bizarre argument to me and I have never heard it before. If you don't want commercial enterprises to know about your bugs, don't release your product under the AGPL. 

Unfortunately, the level of professionalism and customer-focus shown means that I could never recommend Nextcloud. I have torn-down my Nextcloud VM and am working on trialling an alternative solution. 

For what it's worth, I found the server to be quite impressive, the rest not so much. Your Android client is OK but needs a lot more work. Your mistake is to assume that any critique on your product are a critique on the developers, but this is not the case. Nextcloud needs to look at itself and decide what it is... are you are a for-profit company or an open-source charity? Are you employees or volunteers? If you are employees then your job is to serve customers, not insult them, and to take honest feedback on your product on-board and do what you need to do to make it better. That is what it means to be in business.

**Comment by jancborchardt** [comment]
@dothebakedpotato please note @camilasan’s reply and apology above at https://github.com/nextcloud/desktop/issues/1000#issuecomment-490837089 – no need to further inflame this conversation.

**Comment by Germano0** [comment]
I have just tested https://github.com/nextcloud/desktop/pull/1245 and it seems to fix the problem. I have just released a new updated version into [Fedora testing repositories](https://bodhi.fedoraproject.org/updates/?packages=nextcloud-client).

This bug affected also owncloud client 2.5.0 https://github.com/owncloud/client/issues/6804 You may want to check how they fixed it https://github.com/owncloud/client/pull/6806/files

**Comment by bill-mcgonigle** [comment]
paging @ivaradi for Debian (FYI, István).

**Comment by camilasan** [comment]
PR merged.

**Comment by tomharristech** [comment]
I have folders that have been moved around previously that haven't been synced - any ideas on how I can sort this out?

**Comment by tobiasKaminsky** [comment]
Restart the client should help

**Comment by matzbert** [comment]
same bug still occurs. Latest Version 2.5.2git (build 20190319), Windows 7.
Subfolders are synced but without files inside. After stopping the client and restart, everything works.
Seriously a bug in a basic function since January is not solved?????

**Comment by zocker-160** [comment]
I have this same bug as well.

version 2.5.2 on Linux and Windows.

**Comment by NoICE** [comment]
I just installed latest available nextcloud client from Ubuntu PPA (2.5.2git), copied large folder to my $HOME/Nextcloud folder and the contents of the subfolders are not synced.Which version should have this fix implemented? The PR was merged on May, but I take it didn't fix this completely. Issue #1083 mentions this issue, but this issue is already closed. So what is the state of this issue?

**Comment by zocker-160** [comment]
@NoICE this fix is available in the latest version which is v2.5.3
https://github.com/nextcloud/desktop/releases
(changelog does mention this issue)

so your version is not up to date and has not this fix yet

**Comment by NoICE** [comment]
@zocker-160 I just went to Releases here on github and noticed it. My version 2.5.2 is downloaded from https://nextcloud.com/install/#install-clients, so the web is not up-to-date. Will try with 2.5.3 from here. Thank you.

**Comment by jendib** [comment]
I'm still experiencing this issue on 2.6.0 on MacOS. Anyone else?

**Comment by joshua-sterner** [comment]
@jendib I was able to reproduce this on 2.6.0 on Linux for folder structures that were copied. Folder structures that were moved but not copied uploaded correctly. This is weird because the original issue occurred with moved folders, not copied folders. I'll see if I can find out why this broke.

**Comment by jendib** [comment]
In my case it's with moved folders. Only the first level is synced, but subfolders are not. Tell me if I can be of any help.

**Comment by k9t9** [comment]
Have same problem with moved folders, Win7 / Synch client 2.6.0 build 20190927

**Comment by OptimusGREEN** [comment]
This is still an issue as of today with everything up to date.

**Comment by wpoet** [comment]
As per my post above, I tried
`mkdir -p dir1/dir2/dir3/dir4`
In contrast to the above mentioned post, **dir4 is now synced**, although being empty
I have Ubuntu 20.04, with the client Version 2.6.4git, installed via the Ubuntu ppa
`sudo add-apt-repository ppa:nextcloud-devs/client`
The server is from nextcloud-19.0.0.zip.

**Comment by abysso2** [comment]
Hi wpoet,
i also use Ubuntu 20.04 and the nc-client 2.6.4, but here i use the appimage. My server is running nc-version 18.0.5. 

I tested the following:

mkdir -p dir1/dir2/dir3/dir4
touch dir1/dir2/dir3/dir4/test
touch dir1/dir2/dir3/test
touch dir1/dir2/test
touch dir1/test

Everything is synced properly ...

**Comment by DerHOgi** [comment]
Still the same problem with the latest macOS client :(

**Comment by camilasan** [comment]
> Still the same problem with the latest macOS client :( Why is this issue closed???

Could you share your logs after running the client with --logdebug? Including syncing logs and the db files in your Nextcloud configuration folder?  Which folder/file is not synced?

If anyone else is having this issue please do the same.

**Comment by DerHOgi** [comment]
Please check out this bug report and the video. You can also download my testfiles and try it by yourself. Please let me know if you have the same behavior.

https://github.com/nextcloud/desktop/issues/2176

Am Dienstag, 28. Juli 2020 um 16:39 schrieb Camila:

**Comment by claell** [comment]
Since according to the comments, this has not really been fixed, I'll reopen this issue.

**Comment by Lars1963** [comment]
Same problem here with desktop 3.4.1 (without VFS) and nc 23.0.0

**Comment by abysso2** [comment]
I can not reproduce this issue (Client 3.4.1 - NC 22.2.3): 
I created
   Folder 1 (with 2 files)
      Subfolder 2 (with 2 files)
         Subfolder 3 (with 2 files)

Copy / pasted folder 1 to a nextcloud share within windows explorer - all good, no logentries ...

**Comment by Lars1963** [comment]
@abysso2 Yes, wich NC22 it was all good, but with NC**23** it isn't anymore

**Comment by Pas02** [comment]
I can confirm this error with NC23.0.0 and the sync-Client 3.4.1 (Windows) under windowws-10.0.19044

**Comment by OmarWazzan** [comment]
Can confirm - NC23.0.0 and Windows client 3.4.1

**Comment by claell** [comment]
Adding the approved label, so the bot doesn't close this over and over.

**Comment by joshtrichards** [comment]
Reading through this and the associated PRs I believe there may have been some confusion because of two groups of people impacted my similar behavior but with somewhat different solutions. 

One group was addressed via #1245 (which made its way into >= v2.5.3).

The other group of people experience a similar situation that was addressed in #2439 (which made its way into >= v3.0.2). 

This thread was quiet after that, but reopened two years later. 

If you're experiencing anything like this today, it's for a different reason. Please create a dedicated issue for it so that we have accurate details about your situation (and so that it doesn't get lost here in a thread about two matters that were long ago addressed).

For the above reasons - and because this thread hasn't seen any new traffic >2 years - I'm closing this out. 

Again, do not hesitate to create a dedicated issue for your situation though if you're experiencing anything like this today with up-to-date client versions.
